packages/back-end/src/events/handlers/slack/handleFeatureEventForSlack.ts

```diff
@@ -0,0 +1,312 @@
+  eventId: string;
+  featureEvent:
+    | FeatureCreatedNotificationEvent
+    | FeatureUpdatedNotificationEvent
+    | FeatureDeletedNotificationEvent;
+};
+export const handleFeatureEventForSlack = async ({
+  organizationId,
+  eventId,
+  featureEvent,
+}: HandleFeatureEventOptions) => {
+  // Build filtering query and get relevant SlackIntegration records
+  const tags = getTagsForFeatureEvent(featureEvent);
+  const environments = getEnvironmentsForFeatureEvent(featureEvent);
+  const projects = getProjectsForFeatureEvent(featureEvent);
+
+  const slackIntegrations = await getSlackIntegrationsForFilters({
+    organizationId,
+    eventName: featureEvent.event,
+    tags,
+    environments,
+    projects,
+  });
+
+  slackIntegrations?.forEach((slackIntegration) => {
+    // Build the Slack message for the given event
+    const slackMessage = buildSlackMessageForEvent({
+      eventId,
+      slackIntegration,
+      featureEvent,
+    });
+
+    sendSlackMessage(slackMessage, slackIntegration.slackIncomingWebHook).then(
+      (isSuccessful) => {
+        if (!isSuccessful) {
+          logger.warn("Failed to notify for Slack integration", {
+            id: slackIntegration.id,
+          });
+        }
+      }
+    );
+  });
+};
+
+/**
+ * Gets current and previous projects
+ * @param featureEvent
+ */
+const getProjectsForFeatureEvent = (
+  featureEvent:
+    | FeatureCreatedNotificationEvent
+    | FeatureUpdatedNotificationEvent
+    | FeatureDeletedNotificationEvent
+): string[] => {
+  switch (featureEvent.event) {
+    case "feature.created":
+      return featureEvent.data.current.project
+        ? [featureEvent.data.current.project]
+        : [];
+
+    case "feature.updated":
+      return uniq(
+        (featureEvent.data.current.project
+          ? [featureEvent.data.current.project]
+          : []
+        ).concat(
+          featureEvent.data.previous.project
+            ? [featureEvent.data.previous.project]
+            : []
+        )
+      );
+
+    case "feature.deleted":
+      return featureEvent.data.previous.project
+        ? [featureEvent.data.previous.project]
+        : [];
+  }
+};
+
+/**
+ * Gets all current and previous tags for the event
+ * @param featureEvent
+ */
+const getTagsForFeatureEvent = (
+  featureEvent:
+    | FeatureCreatedNotificationEvent
+    | FeatureUpdatedNotificationEvent
+    | FeatureDeletedNotificationEvent
+): string[] => {
+  switch (featureEvent.event) {
+    case "feature.created":
+      return featureEvent.data.current.tags || [];
+
+    case "feature.updated":
+      return uniq(
+        (featureEvent.data.current.tags || []).concat(
+          featureEvent.data.previous.tags || []
+        )
+      );
+
+    case "feature.deleted":
+      return featureEvent.data.previous.tags || [];
+  }
+};
+
+/**
+ * The relevant environments are any environments that are either currently enabled
+ * or were previously enabled
+ * @param featureEvent
+ */
```

Comment:
I'm making the assumption here that if you want to filter using "production" then all relevant events where the production environment is enabled (e.g. for created, deleted and updated events) as well as relevant events for if an environment was toggled on or off (e.g. for updated events in the case we toggle the environment from on to off).

Reply:
For the updated event specifically, we might want to change this logic a bit to only consider a few specific fields when looking for changes, otherwise it might get really noisy in Slack.  Look in the function `getSDKPayloadKeysByDiff` for an example of this.

If someone says they only want to be notified of production changes, then I would think they only care about changes which are actually going to affect how the feature behaves in production.  So updating the markdown description doesn't matter.  Neither does making a bunch of draft changes that aren't published yet.

Reply:
I've made some changes to perform the filtering outside of MongoDB, and outside of the model, since this is business logic related to Slack-specific events we care to notify for.

Please see the `handleFeatureEventForSlack` method and helper methods, mainly `filterForRelevance` was added. 

MongoDB filtering:

- only relevant tags
- only relevant projects
- only relevant events
- **all** environments

Since environments are more complicated, I am using similar logic from `getSDKPayloadKeysByDiff` to figure out which properties of a feature are relevant (i.e. all-environment properties), as well as the changed environments. Then I'm checking those changed environments with the Slack integration's environment filters.

Let me know if we need further changes.